// angular-ng-autocomplete/bundles/autocomplete-lib.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('autocomplete-lib', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/forms', '@angular/common'], factory) :
    (factory((global['autocomplete-lib'] = {}),global.ng.core,global.rxjs,global.rxjs.operators,global.ng.forms,global.ng.common));
}(this, (function (exports,i0,rxjs,operators,forms,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AutocompleteLibService = (function () {
        function AutocompleteLibService() {
        }
        AutocompleteLibService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] },
        ];
        /** @nocollapse */
        AutocompleteLibService.ctorParameters = function () { return []; };
        /** @nocollapse */ AutocompleteLibService.ngInjectableDef = i0.defineInjectable({ factory: function AutocompleteLibService_Factory() { return new AutocompleteLibService(); }, token: AutocompleteLibService, providedIn: "root" });
        return AutocompleteLibService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AutocompleteLibComponent = (function () {
        function AutocompleteLibComponent() {
        }
        /**
         * @return {?}
         */
        AutocompleteLibComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
            };
        AutocompleteLibComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'ng-autocomplete-lib',
                        template: "\n    <p>\n      autocomplete-lib works!\n    </p>\n  ",
                        styles: []
                    },] },
        ];
        /** @nocollapse */
        AutocompleteLibComponent.ctorParameters = function () { return []; };
        return AutocompleteLibComponent;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** *
     * Keyboard events
      @type {?} */
    var isArrowUp = function (keyCode) { return keyCode === 38; };
    /** @type {?} */
    var isArrowDown = function (keyCode) { return keyCode === 40; };
    /** @type {?} */
    var isArrowUpDown = function (keyCode) { return isArrowUp(keyCode) || isArrowDown(keyCode); };
    /** @type {?} */
    var isEnter = function (keyCode) { return keyCode === 13; };
    /** @type {?} */
    var isBackspace = function (keyCode) { return keyCode === 8; };
    /** @type {?} */
    var isDelete = function (keyCode) { return keyCode === 46; };
    /** @type {?} */
    var isESC = function (keyCode) { return keyCode === 27; };
    /** @type {?} */
    var isTab = function (keyCode) { return keyCode === 9; };
    var AutocompleteComponent = (function () {
        function AutocompleteComponent(myElemenetRef) {
            this.query = '';
            this.filteredList = [];
            this.historyList = [];
            this.isHistoryListVisible = true;
            this.toHighlight = '';
            this.notFound = false;
            /**
             * Data of items list.
             * It can be array of strings or array of objects.
             */
            this.data = [];
            this.placeHolder = '';
            /**
             * Heading text of history list.
             * If it is null then history heading is hidden.
             */
            this.historyHeading = 'Recently selected';
            this.historyListMaxNumber = 15;
            this.notFoundText = 'Not found';
            /**
             * Event that is emitted whenever an item from the list is selected.
             */
            this.selected = new i0.EventEmitter();
            /**
             * Event that is emitted whenever an input is changed.
             */
            this.inputChanged = new i0.EventEmitter();
            /**
             * Event that is emitted whenever an input is focused.
             */
            this.inputFocused = new i0.EventEmitter();
            /**
             * Event that is emitted when the autocomplete panel is opened.
             */
            this.opened = new i0.EventEmitter();
            /**
             * Event that is emitted when the autocomplete panel is closed.
             */
            this.closed = new i0.EventEmitter();
            this.elementRef = myElemenetRef;
            this.selectedIdx = -1;
        }
        /**
         * @return {?}
         */
        AutocompleteComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.initEventStream();
                this.setInitialValue(this.initialValue);
            };
        /**
         * Set initial value
         * @param {?} value
         * @return {?}
         */
        AutocompleteComponent.prototype.setInitialValue = /**
         * Set initial value
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (this.initialValue) {
                    this.select(value);
                }
            };
        /**
         * Update search results
         */
        /**
         * Update search results
         * @param {?} changes
         * @return {?}
         */
        AutocompleteComponent.prototype.ngOnChanges = /**
         * Update search results
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (changes && changes["data"] &&
                    Array.isArray(changes["data"].currentValue)) {
                    if (!changes["data"].firstChange && this.isFocused) {
                        this.open();
                    }
                }
            };
        /**
         * Filter data
         * @return {?}
         */
        AutocompleteComponent.prototype.filterList = /**
         * Filter data
         * @return {?}
         */
            function () {
                var _this = this;
                this.initSearchHistory();
                if (this.query != null && this.data) {
                    this.toHighlight = this.query;
                    this.filteredList = this.data.filter(function (item) {
                        if (typeof item === 'string') {
                            // string logic, check equality of strings
                            return item.toLowerCase().indexOf(_this.query.toLowerCase()) > -1;
                        }
                        else if (typeof item === 'object' && item.constructor === Object) {
                            // object logic, check property equality
                            return item[_this.searchKeyword].toLowerCase().indexOf(_this.query.toLowerCase()) > -1;
                        }
                    });
                }
                else {
                    this.notFound = false;
                }
            };
        /**
         * Check type of item in the list.
         * @param item
         */
        /**
         * Check type of item in the list.
         * @param {?} item
         * @return {?}
         */
        AutocompleteComponent.prototype.isType = /**
         * Check type of item in the list.
         * @param {?} item
         * @return {?}
         */
            function (item) {
                return typeof item === 'string';
            };
        /**
         * Select item in the list.
         * @param {?} item
         * @return {?}
         */
        AutocompleteComponent.prototype.select = /**
         * Select item in the list.
         * @param {?} item
         * @return {?}
         */
            function (item) {
                var _this = this;
                this.query = !this.isType(item) ? item[this.searchKeyword] : item;
                this.selected.emit(item);
                if (this.initialValue) {
                    /** @type {?} */
                    var history_1 = window.localStorage.getItem("" + this.historyIdentifier);
                    if (history_1) {
                        /** @type {?} */
                        var existingHistory = JSON.parse(localStorage["" + this.historyIdentifier]);
                        if (!(existingHistory instanceof Array))
                            existingHistory = [];
                        // check if selected item exists in existingHistory
                        if (!existingHistory.some(function (existingItem) {
                            return !_this.isType(existingItem)
                                ? existingItem[_this.searchKeyword] == item[_this.searchKeyword] : existingItem == item;
                        })) {
                            existingHistory.unshift(item);
                            localStorage.setItem("" + this.historyIdentifier, JSON.stringify(existingHistory));
                            // check if items don't exceed max allowed number
                            if (existingHistory.length >= this.historyListMaxNumber) {
                                existingHistory.splice(existingHistory.length - 1, 1);
                                localStorage.setItem("" + this.historyIdentifier, JSON.stringify(existingHistory));
                            }
                        }
                    }
                    else {
                        this.saveHistory(item);
                    }
                }
                else {
                    this.saveHistory(item);
                }
                this.close();
            };
        /**
         * Document click
         * @param {?} e event
         * @return {?}
         */
        AutocompleteComponent.prototype.handleClick = /**
         * Document click
         * @param {?} e event
         * @return {?}
         */
            function (e) {
                /** @type {?} */
                var clickedComponent = e.target;
                /** @type {?} */
                var inside = false;
                do {
                    if (clickedComponent === this.elementRef.nativeElement) {
                        inside = true;
                        if (this.filteredList.length) {
                            this.open();
                        }
                    }
                    clickedComponent = clickedComponent.parentNode;
                } while (clickedComponent);
                if (!inside) {
                    this.close();
                }
            };
        /**
         * Remove search query
         * @return {?}
         */
        AutocompleteComponent.prototype.remove = /**
         * Remove search query
         * @return {?}
         */
            function () {
                this.query = '';
                this.close();
            };
        /**
         * Initialize historyList search
         */
        /**
         * Initialize historyList search
         * @return {?}
         */
        AutocompleteComponent.prototype.initSearchHistory = /**
         * Initialize historyList search
         * @return {?}
         */
            function () {
                this.isHistoryListVisible = false;
                if (this.historyIdentifier && !this.query) {
                    /** @type {?} */
                    var history_2 = window.localStorage.getItem("" + this.historyIdentifier);
                    if (history_2) {
                        this.isHistoryListVisible = true;
                        this.filteredList = [];
                        this.historyList = history_2 ? JSON.parse(history_2) : [];
                    }
                    else {
                        this.isHistoryListVisible = false;
                    }
                }
                else {
                    this.isHistoryListVisible = false;
                }
            };
        /**
         * @return {?}
         */
        AutocompleteComponent.prototype.open = /**
         * @return {?}
         */
            function () {
                if (this.isOpen || this.isOpen && !this.isLoading) {
                    return;
                }
                // If data exists
                if (this.data && this.data.length) {
                    this.filterList();
                    this.opened.emit();
                    this.isOpen = true;
                }
            };
        /**
         * @return {?}
         */
        AutocompleteComponent.prototype.close = /**
         * @return {?}
         */
            function () {
                if (!this.isOpen) {
                    return;
                }
                this.filteredList = [];
                this.selectedIdx = -1;
                this.notFound = false;
                this.isHistoryListVisible = false;
                this.isFocused = false;
                this.closed.emit();
                this.isOpen = false;
            };
        /**
         * @param {?} e
         * @return {?}
         */
        AutocompleteComponent.prototype.focus = /**
         * @param {?} e
         * @return {?}
         */
            function (e) {
                //this.searchInput.nativeElement.focus();
                if (this.isFocused) {
                    return;
                }
                // if data exists then open
                if (this.data && this.data.length) {
                    this.open();
                }
                this.inputFocused.emit(e);
                this.isFocused = true;
            };
        /**
         * Initialize keyboard events
         */
        /**
         * Initialize keyboard events
         * @return {?}
         */
        AutocompleteComponent.prototype.initEventStream = /**
         * Initialize keyboard events
         * @return {?}
         */
            function () {
                this.inputKeyUp$ = rxjs.fromEvent(this.searchInput.nativeElement, 'keyup').pipe(operators.map(function (e) { return e; }));
                this.inputKeyDown$ = rxjs.fromEvent(this.searchInput.nativeElement, 'keydown').pipe(operators.map(function (e) { return e; }));
                this.listenEventStream();
            };
        /**
         * Listen keyboard events
         */
        /**
         * Listen keyboard events
         * @return {?}
         */
        AutocompleteComponent.prototype.listenEventStream = /**
         * Listen keyboard events
         * @return {?}
         */
            function () {
                var _this = this;
                // key up event
                this.inputKeyUp$
                    .pipe(operators.filter(function (e) {
                    return !isArrowUpDown(e.keyCode) &&
                        !isEnter(e.keyCode) &&
                        !isESC(e.keyCode) &&
                        !isTab(e.keyCode);
                }), operators.debounceTime(this.debounceTime)).subscribe(function (e) {
                    _this.onKeyUp(e);
                });
                // cursor up & down
                this.inputKeyDown$.pipe(operators.filter(function (e) { return isArrowUpDown(e.keyCode); })).subscribe(function (e) {
                    e.preventDefault();
                    _this.onFocusNextItem(e);
                });
                // enter
                this.inputKeyUp$.pipe(operators.filter(function (e) { return isEnter(e.keyCode); })).subscribe(function (e) {
                    _this.onHandleEnter();
                });
                // ESC
                this.inputKeyUp$.pipe(operators.filter(function (e) { return isESC(e.keyCode); }, operators.debounceTime(100))).subscribe(function (e) {
                    _this.onEsc();
                });
                // delete
                this.inputKeyDown$.pipe(operators.filter(function (e) { return isBackspace(e.keyCode) || isDelete(e.keyCode); })).subscribe(function (e) {
                    _this.onDelete();
                });
            };
        /**
         * on keyup == when input changed
         * @param e event
         */
        /**
         * on keyup == when input changed
         * @param {?} e event
         * @return {?}
         */
        AutocompleteComponent.prototype.onKeyUp = /**
         * on keyup == when input changed
         * @param {?} e event
         * @return {?}
         */
            function (e) {
                this.notFound = false; // search results are unknown while typing
                if (!this.query) {
                    this.notFound = false;
                }
                this.inputChanged.emit(e.target.value);
                this.filterList();
                // If no results found
                if (!this.filteredList.length) {
                    this.notFoundText ? this.notFound = true : this.notFound = false;
                }
            };
        /**
         * Keyboard arrow top and arrow bottom input
         * @param e event
         */
        /**
         * Keyboard arrow top and arrow bottom input
         * @param {?} e event
         * @return {?}
         */
        AutocompleteComponent.prototype.onFocusNextItem = /**
         * Keyboard arrow top and arrow bottom input
         * @param {?} e event
         * @return {?}
         */
            function (e) {
                // move arrow up and down on filteredList or historyList
                if (!this.historyList.length || !this.isHistoryListVisible) {
                    // filteredList
                    if (e.code === 'ArrowDown' && this.selectedIdx < this.filteredList.length - 1) {
                        this.selectedIdx++;
                    }
                    else if (e.code === 'ArrowUp' && this.selectedIdx > 0) {
                        this.selectedIdx--;
                    }
                }
                else {
                    // historyList
                    if (e.code === 'ArrowDown' && this.selectedIdx < this.historyList.length - 1) {
                        this.selectedIdx++;
                    }
                    else if (e.code === 'ArrowUp' && this.selectedIdx > 0) {
                        this.selectedIdx--;
                    }
                }
            };
        /**
         * Select item on enter click
         */
        /**
         * Select item on enter click
         * @return {?}
         */
        AutocompleteComponent.prototype.onHandleEnter = /**
         * Select item on enter click
         * @return {?}
         */
            function () {
                // click enter to choose item from filteredList or historyList
                if (this.selectedIdx > -1) {
                    if (!this.historyList.length || !this.isHistoryListVisible) {
                        // filteredList
                        this.query = !this.isType(this.filteredList[this.selectedIdx])
                            ? this.filteredList[this.selectedIdx][this.searchKeyword]
                            : this.filteredList[this.selectedIdx];
                        this.saveHistory(this.filteredList[this.selectedIdx]);
                        this.select(this.filteredList[this.selectedIdx]);
                    }
                    else {
                        // historyList
                        this.query = !this.isType(this.historyList[this.selectedIdx])
                            ? this.historyList[this.selectedIdx][this.searchKeyword]
                            : this.historyList[this.selectedIdx];
                        this.saveHistory(this.historyList[this.selectedIdx]);
                        this.select(this.historyList[this.selectedIdx]);
                    }
                }
                this.isHistoryListVisible = false;
                this.close();
            };
        /**
         * Esc click
         */
        /**
         * Esc click
         * @return {?}
         */
        AutocompleteComponent.prototype.onEsc = /**
         * Esc click
         * @return {?}
         */
            function () {
                this.searchInput.nativeElement.blur();
                this.close();
            };
        /**
         * Delete click
         */
        /**
         * Delete click
         * @return {?}
         */
        AutocompleteComponent.prototype.onDelete = /**
         * Delete click
         * @return {?}
         */
            function () {
                //console.log('delete');
            };
        /**
         * Select item to save in localStorage
         * @param selected
         */
        /**
         * Select item to save in localStorage
         * @param {?} selected
         * @return {?}
         */
        AutocompleteComponent.prototype.saveHistory = /**
         * Select item to save in localStorage
         * @param {?} selected
         * @return {?}
         */
            function (selected) {
                var _this = this;
                if (this.historyIdentifier) {
                    // check if selected item exists in historyList
                    if (!this.historyList.some(function (item) {
                        return !_this.isType(item)
                            ? item[_this.searchKeyword] == selected[_this.searchKeyword] : item == selected;
                    })) {
                        this.saveHistoryToLocalStorage(__spread([selected], this.historyList));
                        // check if items don't exceed max allowed number
                        if (this.historyList.length >= this.historyListMaxNumber) {
                            this.historyList.splice(this.historyList.length - 1, 1);
                            this.saveHistoryToLocalStorage(__spread([selected], this.historyList));
                        }
                    }
                }
            };
        /**
         * Save item in localStorage
         * @param selected
         */
        /**
         * Save item in localStorage
         * @param {?} selected
         * @return {?}
         */
        AutocompleteComponent.prototype.saveHistoryToLocalStorage = /**
         * Save item in localStorage
         * @param {?} selected
         * @return {?}
         */
            function (selected) {
                window.localStorage.setItem("" + this.historyIdentifier, JSON.stringify(selected));
            };
        /**
         * Remove item from localStorage
         * @param index
         * @param e event
         */
        /**
         * Remove item from localStorage
         * @param {?} index
         * @param {?} e event
         * @return {?}
         */
        AutocompleteComponent.prototype.removeHistoryItem = /**
         * Remove item from localStorage
         * @param {?} index
         * @param {?} e event
         * @return {?}
         */
            function (index, e) {
                e.stopPropagation();
                this.historyList = this.historyList.filter(function (v, i) { return i !== index; });
                this.saveHistoryToLocalStorage(this.historyList);
                if (this.historyList.length == 0) {
                    window.localStorage.removeItem("" + this.historyIdentifier);
                    this.filterList();
                }
            };
        /**
         * Reset localStorage
         * @param e event
         */
        /**
         * Reset localStorage
         * @param {?} e event
         * @return {?}
         */
        AutocompleteComponent.prototype.resetHistoryList = /**
         * Reset localStorage
         * @param {?} e event
         * @return {?}
         */
            function (e) {
                e.stopPropagation();
                this.historyList = [];
                window.localStorage.removeItem("" + this.historyIdentifier);
                this.filterList();
            };
        AutocompleteComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'ng-autocomplete',
                        template: "<div class=\"autocomplete-container\">\n  <div class=\"input-container\">\n    <input #searchInput type=\"text\" placeholder={{placeHolder}}\n           [(ngModel)]=query\n           (focus)=focus($event)>\n    <div class=\"x\" *ngIf=\"query && !isLoading\" (click)=\"remove()\">\n      <i class=\"material-icons\">close</i>\n    </div>\n    <!--Loading mask-->\n    <div class=\"sk-fading-circle\" *ngIf=\"isLoading\">\n      <div class=\"sk-circle1 sk-circle\"></div>\n      <div class=\"sk-circle2 sk-circle\"></div>\n      <div class=\"sk-circle3 sk-circle\"></div>\n      <div class=\"sk-circle4 sk-circle\"></div>\n      <div class=\"sk-circle5 sk-circle\"></div>\n      <div class=\"sk-circle6 sk-circle\"></div>\n      <div class=\"sk-circle7 sk-circle\"></div>\n      <div class=\"sk-circle8 sk-circle\"></div>\n      <div class=\"sk-circle9 sk-circle\"></div>\n      <div class=\"sk-circle10 sk-circle\"></div>\n      <div class=\"sk-circle11 sk-circle\"></div>\n      <div class=\"sk-circle12 sk-circle\"></div>\n    </div>\n  </div>\n\n  <!--FilteredList items-->\n  <div class=\"suggestions-container\"\n       [ngClass]=\"{ 'is-hidden': isHistoryListVisible, 'is-visible': !isHistoryListVisible}\">\n    <ul>\n      <li *ngFor=\"let item of filteredList; let idx = index\">\n        <!--string logic-->\n        <div [class.complete-selected]=\"idx === selectedIdx\" *ngIf='isType(item)'\n             (click)=\"select(item)\">\n          <ng-container\n            *ngTemplateOutlet=\"itemTemplate;  context: { $implicit: item | highlight: toHighlight }\">\n          </ng-container>\n        </div>\n        <!--object logic-->\n        <div [class.complete-selected]=\"idx === selectedIdx\" *ngIf='!isType(item)'\n             (click)=\"select(item)\">\n          <ng-container\n            *ngTemplateOutlet=\"itemTemplate; context: { $implicit: item | highlight: toHighlight : searchKeyword }\">\n          </ng-container>\n        </div>\n\n      </li>\n    </ul>\n  </div>\n\n  <!--HistoryList items-->\n  <div class=\"suggestions-container\"\n       [ngClass]=\"{ 'is-hidden': !isHistoryListVisible, 'is-visible': isHistoryListVisible}\">\n    <!--HistoryList heading-->\n    <div class=\"history-heading\" *ngIf=\"historyList.length > 0 && historyHeading\">\n      <div class=\"text\">{{historyHeading}}</div>\n      <div class=\"x\" (click)=\"resetHistoryList($event)\">\n        <i class=\"material-icons\">delete</i>\n      </div>\n    </div>\n\n    <ul *ngFor=\"let item of historyList; let idx = index\">\n      <li [class.complete-selected]=\"idx === selectedIdx\">\n        <!--string logic-->\n        <div *ngIf='isType(item)' (click)=\"select(item)\">\n          <ng-container\n            *ngTemplateOutlet=\"itemTemplate;  context: { $implicit: item }\">\n          </ng-container>\n        </div>\n        <!--object logic-->\n        <div *ngIf='!isType(item)' (click)=\"select(item)\">\n          <ng-container\n            *ngTemplateOutlet=\"itemTemplate; context: { $implicit: item }\">\n          </ng-container>\n        </div>\n        <div class=\"x\" (click)=\"removeHistoryItem(idx, $event)\">\n          <i class=\"material-icons\">close</i>\n        </div>\n      </li>\n    </ul>\n  </div>\n\n  <!--Not found-->\n  <div class=\"not-found\" *ngIf=\"isLoading ? !isLoading && notFound : notFound\">\n    <ng-container\n      *ngTemplateOutlet=\"notFoundTemplate;  context: { $implicit: notFoundText  }\">\n    </ng-container>\n  </div>\n</div>\n",
                        styles: ["@import url(https://fonts.googleapis.com/icon?family=Material+Icons);.ng-autocomplete{width:600px}.autocomplete-container{box-shadow:0 1px 3px 0 rgba(0,0,0,.2),0 1px 1px 0 rgba(0,0,0,.14),0 2px 1px -1px rgba(0,0,0,.12);position:relative;overflow:visible;height:40px}.autocomplete-container .input-container input{font-size:14px;box-sizing:border-box;border:none;box-shadow:none;outline:0;background:0 0;color:rgba(0,0,0,.87);width:100%;padding:0 15px;line-height:40px;height:40px}.autocomplete-container .input-container .x{position:absolute;right:10px;margin:auto;cursor:pointer;top:50%;-webkit-transform:translateY(-50%);transform:translateY(-50%)}.autocomplete-container .input-container .x i{color:rgba(0,0,0,.54);font-size:22px;vertical-align:middle}.autocomplete-container .suggestions-container{border:1px solid #f1f1f1;position:absolute;width:100%;background:#fff;height:auto;box-sizing:border-box;max-height:240px;overflow-y:auto}.autocomplete-container .suggestions-container ul{padding:0;margin:0}.autocomplete-container .suggestions-container ul li{position:relative;list-style:none;padding:0;margin:0;cursor:pointer;overflow:hidden}.autocomplete-container .suggestions-container ul li a{padding:5px 15px;display:block;text-decoration:none;cursor:pointer;color:rgba(0,0,0,.87);font-size:15px}.autocomplete-container .suggestions-container .complete-selected,.autocomplete-container .suggestions-container ul li:hover{background-color:rgba(158,158,158,.18)}.autocomplete-container .suggestions-container .history-heading{position:relative;padding:0 .75em}.autocomplete-container .suggestions-container .history-heading .text{padding:.3em 0;font-size:.85em;line-height:1.4;border-bottom:1px solid rgba(230,230,230,.7)}.autocomplete-container .suggestions-container .x{position:absolute;right:10px;margin:auto;cursor:pointer;top:50%;-webkit-transform:translateY(-50%);transform:translateY(-50%)}.autocomplete-container .suggestions-container .x i{color:rgba(0,0,0,.54);font-size:18px;vertical-align:middle}.autocomplete-container .suggestions-container.is-hidden{visibility:hidden}.autocomplete-container .suggestions-container.is-visible{visibility:visible}.autocomplete-container .not-found{padding:0 .75em;border:1px solid #f1f1f1;background:#fff}.autocomplete-container .not-found div{padding:.4em 0;font-size:.95em;line-height:1.4;border-bottom:1px solid rgba(230,230,230,.7)}.highlight{font-weight:700}.sk-fading-circle{width:20px;height:20px;position:absolute;right:10px;top:0;bottom:0;margin:auto}.sk-fading-circle .sk-circle{width:100%;height:100%;position:absolute;left:0;top:0}.sk-fading-circle .sk-circle:before{content:'';display:block;margin:0 auto;width:15%;height:15%;background-color:#333;border-radius:100%;-webkit-animation:1.2s ease-in-out infinite both sk-circleFadeDelay;animation:1.2s ease-in-out infinite both sk-circleFadeDelay}.sk-fading-circle .sk-circle2{-webkit-transform:rotate(30deg);transform:rotate(30deg)}.sk-fading-circle .sk-circle3{-webkit-transform:rotate(60deg);transform:rotate(60deg)}.sk-fading-circle .sk-circle4{-webkit-transform:rotate(90deg);transform:rotate(90deg)}.sk-fading-circle .sk-circle5{-webkit-transform:rotate(120deg);transform:rotate(120deg)}.sk-fading-circle .sk-circle6{-webkit-transform:rotate(150deg);transform:rotate(150deg)}.sk-fading-circle .sk-circle7{-webkit-transform:rotate(180deg);transform:rotate(180deg)}.sk-fading-circle .sk-circle8{-webkit-transform:rotate(210deg);transform:rotate(210deg)}.sk-fading-circle .sk-circle9{-webkit-transform:rotate(240deg);transform:rotate(240deg)}.sk-fading-circle .sk-circle10{-webkit-transform:rotate(270deg);transform:rotate(270deg)}.sk-fading-circle .sk-circle11{-webkit-transform:rotate(300deg);transform:rotate(300deg)}.sk-fading-circle .sk-circle12{-webkit-transform:rotate(330deg);transform:rotate(330deg)}.sk-fading-circle .sk-circle2:before{-webkit-animation-delay:-1.1s;animation-delay:-1.1s}.sk-fading-circle .sk-circle3:before{-webkit-animation-delay:-1s;animation-delay:-1s}.sk-fading-circle .sk-circle4:before{-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-fading-circle .sk-circle5:before{-webkit-animation-delay:-.8s;animation-delay:-.8s}.sk-fading-circle .sk-circle6:before{-webkit-animation-delay:-.7s;animation-delay:-.7s}.sk-fading-circle .sk-circle7:before{-webkit-animation-delay:-.6s;animation-delay:-.6s}.sk-fading-circle .sk-circle8:before{-webkit-animation-delay:-.5s;animation-delay:-.5s}.sk-fading-circle .sk-circle9:before{-webkit-animation-delay:-.4s;animation-delay:-.4s}.sk-fading-circle .sk-circle10:before{-webkit-animation-delay:-.3s;animation-delay:-.3s}.sk-fading-circle .sk-circle11:before{-webkit-animation-delay:-.2s;animation-delay:-.2s}.sk-fading-circle .sk-circle12:before{-webkit-animation-delay:-.1s;animation-delay:-.1s}@-webkit-keyframes sk-circleFadeDelay{0%,100%,39%{opacity:0}40%{opacity:1}}@keyframes sk-circleFadeDelay{0%,100%,39%{opacity:0}40%{opacity:1}}"],
                        encapsulation: i0.ViewEncapsulation.None,
                        host: {
                            '(document:click)': 'handleClick($event)',
                            'class': 'ng-autocomplete'
                        },
                    },] },
        ];
        /** @nocollapse */
        AutocompleteComponent.ctorParameters = function () {
            return [
                { type: i0.ElementRef }
            ];
        };
        AutocompleteComponent.propDecorators = {
            searchInput: [{ type: i0.ViewChild, args: ['searchInput',] }],
            data: [{ type: i0.Input }],
            searchKeyword: [{ type: i0.Input }],
            placeHolder: [{ type: i0.Input }],
            historyIdentifier: [{ type: i0.Input }],
            historyHeading: [{ type: i0.Input }],
            historyListMaxNumber: [{ type: i0.Input }],
            notFoundText: [{ type: i0.Input }],
            isLoading: [{ type: i0.Input }],
            debounceTime: [{ type: i0.Input }],
            initialValue: [{ type: i0.Input }],
            selected: [{ type: i0.Output }],
            inputChanged: [{ type: i0.Output }],
            inputFocused: [{ type: i0.Output }],
            opened: [{ type: i0.Output }],
            closed: [{ type: i0.Output }],
            itemTemplate: [{ type: i0.ContentChild, args: [i0.TemplateRef,] }, { type: i0.Input }],
            notFoundTemplate: [{ type: i0.Input }]
        };
        return AutocompleteComponent;
    }());
    var HighlightPipe = (function () {
        function HighlightPipe() {
        }
        /**
         * @param {?} text
         * @param {?} search
         * @param {?=} searchKeyword
         * @return {?}
         */
        HighlightPipe.prototype.transform = /**
         * @param {?} text
         * @param {?} search
         * @param {?=} searchKeyword
         * @return {?}
         */
            function (text, search, searchKeyword) {
                /** @type {?} */
                var pattern = search.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
                pattern = pattern.split(' ').filter(function (t) {
                    return t.length > 0;
                }).join('|');
                /** @type {?} */
                var regex = new RegExp(pattern, 'gi');
                if (!search) {
                    return text;
                }
                if (searchKeyword) {
                    /** @type {?} */
                    var name_1 = text[searchKeyword].replace(regex, function (match) { return "<b>" + match + "</b>"; });
                    return __assign({}, text, { name: name_1 });
                }
                else {
                    return search ? text.replace(regex, function (match) { return "<b>" + match + "</b>"; }) : text;
                }
            };
        HighlightPipe.decorators = [
            { type: i0.Pipe, args: [{ name: 'highlight' },] },
        ];
        return HighlightPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AutocompleteLibModule = (function () {
        function AutocompleteLibModule() {
        }
        AutocompleteLibModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule
                        ],
                        declarations: [AutocompleteLibComponent, AutocompleteComponent, HighlightPipe],
                        exports: [AutocompleteLibComponent, AutocompleteComponent]
                    },] },
        ];
        return AutocompleteLibModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.AutocompleteLibService = AutocompleteLibService;
    exports.AutocompleteLibComponent = AutocompleteLibComponent;
    exports.AutocompleteLibModule = AutocompleteLibModule;
    exports.AutocompleteComponent = AutocompleteComponent;
    exports.HighlightPipe = HighlightPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
